File: intellij/src/main/kotlin/dev/nylo/plugin/screenshots/project/RouterParser.kt
package dev.nylo.plugin.screenshots.project

import com.intellij.openapi.project.Project
import dev.nylo.plugin.screenshots.model.NyloPage
import java.io.File

/**
 * Extracts the app's pages from `lib/routes/router.dart`.
 *
 * The router lists routes as `router.add(XxxPage.path)` (optionally chained with
 * `.authenticatedRoute()` / `.previewRoute()`). Each `XxxPage` declares its real
 * route string in its own source as `static RouteView path = ("/x", ...)`, so we
 * additionally scan `lib/` for those declarations to resolve class -> route.
 */
object RouterParser {
    private val ADD_LINE = Regex("""router\s*\.\s*add\s*\(\s*(\w+)\s*\.\s*path""")
    private val AUTH = Regex("""\.authenticatedRoute\s*\(""")
    private val CLASS_DECL = Regex("""class\s+(\w+)""")
    private val PATH_DECL = Regex("""static\s+RouteView\s+path\s*=\s*\(\s*["']([^"']+)["']""")

    fun parse(project: Project): List<NyloPage> {
        val base = project.basePath?.let(::File) ?: return emptyList()
        val routerFile = File(base, "lib/routes/router.dart")
        if (!routerFile.isFile) return emptyList()

        val classToRoute = scanPageRoutes(File(base, "lib"))
        val pages = LinkedHashMap<String, NyloPage>()
        routerFile.readText().lineSequence().forEach { raw ->
            val line = raw.trim()
            if (line.startsWith("//")) return@forEach
            val match = ADD_LINE.find(line) ?: return@forEach
            val className = match.groupValues[1]
            if (pages.containsKey(className)) return@forEach
            val resolved = classToRoute[className]
            pages[className] = NyloPage(
                className = className,
                route = resolved ?: deriveRoute(className),
                displayName = displayName(className),
                authenticated = AUTH.containsMatchIn(line),
                routeResolved = resolved != null,
            )
        }
        return pages.values.toList()
    }

    /** Map `XxxPage` -> "/x" by scanning every Dart file that declares a `RouteView path`. */
    private fun scanPageRoutes(libDir: File): Map<String, String> {
        if (!libDir.isDirectory) return emptyMap()
        val out = HashMap<String, String>()
        libDir.walkTopDown()
            .filter { it.isFile && it.extension == "dart" }
            .forEach { file ->
                val text = runCatching { file.readText() }.getOrNull() ?: return@forEach
                if (!text.contains("RouteView path")) return@forEach
                val classes = CLASS_DECL.findAll(text).toList()
                PATH_DECL.findAll(text).forEach { pathMatch ->
                    val route = pathMatch.groupValues[1]
                    val owningClass = classes.lastOrNull { it.range.first < pathMatch.range.first }
                    owningClass?.groupValues?.get(1)?.let { out.putIfAbsent(it, route) }
                }
            }
        return out
    }

    private fun deriveRoute(className: String): String =
        "/" + className.removeSuffix("Page").replace(Regex("([a-z0-9])([A-Z])"), "$1-$2").lowercase()

    private fun displayName(className: String): String =
        className.removeSuffix("Page").replace(Regex("([a-z0-9])([A-Z])"), "$1 $2").trim()
}

File: vscode/src/core/screenshots/routerParser.ts
import { NyloPage } from './model';

/**
 * Port of `dev.nylo.plugin.screenshots.project.RouterParser` (pure half).
 *
 * The router lists routes as `router.add(XxxPage.path)` (optionally chained with
 * `.authenticatedRoute()`). Each `XxxPage` declares its route string in its own source as
 * `static RouteView path = ("/x", ...)`, so {@link scanPageRoutes} resolves class → route from the
 * `lib/**` Dart sources and {@link parseRouter} applies it.
 */

const ADD_LINE = /router\s*\.\s*add\s*\(\s*(\w+)\s*\.\s*path/;
const AUTH = /\.authenticatedRoute\s*\(/;
const CLASS_DECL = /class\s+(\w+)/g;
const PATH_DECL = /static\s+RouteView\s+path\s*=\s*\(\s*["']([^"']+)["']/g;

/** Builds `XxxPage` → `/x` from every Dart source that declares a `RouteView path`. */
export function scanPageRoutes(dartTexts: string[]): Map<string, string> {
  const out = new Map<string, string>();
  for (const text of dartTexts) {
    if (!text.includes('RouteView path')) {
      continue;
    }
    const classes = [...text.matchAll(CLASS_DECL)].map((m) => ({ name: m[1], index: m.index ?? 0 }));
    for (const pathMatch of text.matchAll(PATH_DECL)) {
      const route = pathMatch[1];
      const owning = classes.filter((c) => c.index < (pathMatch.index ?? 0)).pop();
      if (owning && !out.has(owning.name)) {
        out.set(owning.name, route);
      }
    }
  }
  return out;
}

export function parseRouter(routerText: string, classToRoute: Map<string, string>): NyloPage[] {
  const pages = new Map<string, NyloPage>();
  for (const raw of routerText.split('\n')) {
    const line = raw.trim();
    if (line.startsWith('//')) {
      continue;
    }
    const match = ADD_LINE.exec(line);
    if (!match) {
      continue;
    }
    const className = match[1];
    if (pages.has(className)) {
      continue;
    }
    const resolved = classToRoute.get(className);
    pages.set(className, {
      className,
      route: resolved ?? deriveRoute(className),
      displayName: pageDisplayName(className),
      authenticated: AUTH.test(line),
      routeResolved: resolved !== undefined,
    });
  }
  return [...pages.values()];
}

function deriveRoute(className: string): string {
  return '/' + stripSuffix(className, 'Page').replace(/([a-z0-9])([A-Z])/g, '$1-$2').toLowerCase();
}

function pageDisplayName(className: string): string {
  return stripSuffix(className, 'Page').replace(/([a-z0-9])([A-Z])/g, '$1 $2').trim();
}

function stripSuffix(value: string, suffix: string): string {
  return value.endsWith(suffix) ? value.slice(0, value.length - suffix.length) : value;
}
